### DISCUSSION TITLE
Modernise CMAKE compilation

### DISCUSSION THREAD

**Discussion by grimgort** [pelicanmapping/osgearth#1791] [Support Q&A]
Hello,

With the compilation under cmake, I am obliged to fix a large number of painful variables : 
set(OSGMANIPULATOR_LIBRARY ${OSGMANIPULATOR_ROOT}/lib/osgManipulator.lib )
set(OSGMANIPULATOR_LIBRARY_DEBUG ${OSGMANIPULATOR_ROOT}/lib/osgManipulatord.lib )
set(OSGPARTICLE_LIBRARY ${OSGPARTICLE_ROOT}/lib/osgParticle.lib )
set(OSGPARTICLE_LIBRARY_DEBUG ${OSGPARTICLE_ROOT}/lib/osgParticled.lib ) 
set(OSGSHADOW_LIBRARY ${OSGSHADOW_ROOT}/lib/osgShadow.lib )
set(OSGSHADOW_LIBRARY_DEBUG ${OSGSHADOW_ROOT}/lib/osgShadowd.lib )
set(OSGSIM_LIBRARY ${OSGSIM_ROOT}/lib/osgSim.lib )
set(OSGSIM_LIBRARY_DEBUG ${OSGSIM_ROOT}/lib/osgSimd.lib ) 
set(OSGTERRAIN_LIBRARY ${OSGTERRAIN_ROOT} )
set(OSGTERRAIN_LIBRARY_DEBUG ${OSGTERRAIN_ROOT} )
set(OSGTEXT_LIBRARY ${OSGTEXT_ROOT} )
set(OSGTEXT_LIBRARY_DEBUG ${OSGTEXT_ROOT} )
set(OSGUTIL_LIBRARY ${OSGUTIL_ROOT} )
set(OSGUTIL_LIBRARY_DEBUG ${OSGUTIL_ROOT} )
set(OSGVIEWER_LIBRARY ${OSGVIEWER_ROOT} )
set(OSGVIEWER_LIBRARY_DEBUG ${OSGVIEWER_ROOT} )
set(OSGWIDGET_LIBRARY ${OSGWIDGET_ROOT} )
set(OSGWIDGET_LIBRARY_DEBUG ${OSGWIDGET_ROOT} )

In the new versions of CMAKE we can export the configuration of our installed libraries. For the user, you just have to use the OSG_ROOT variable when installing OSG and use find_package(OSG) to link the whole lib to a target.
https://cmake.org/cmake/help/latest/guide/importing-exporting/index.html
a project that works well: https://github.com/pboettch/json-schema-validator

Thanks

**Comment by gwaldron** [comment]
That's right @grimgort. You can also set the new `OSGEARTH_DEPENDENCY_DIR` variable (in osgEarth 3.1+) and it will look there for things like GDAL, CURL, GEOS, etc. 

Thanks for the helpful information :)

**Comment by grimgort** [comment]
thank you for the information. Unfortunately, the different libraries are in different directories (install/gdal, install/zlib,ect...)
So I can't use this variable which adds directly in the cmake path 2 paths.

SET(OSGEARTH_DEPENDENCY_DIR "" CACHE PATH "Folder to search for dependencies (e.g., vcpkg/installed/x64-windows)")

IF(OSGEARTH_DEPENDENCY_DIR)
    LIST(APPEND CMAKE_INCLUDE_PATH ${OSGEARTH_DEPENDENCY_DIR}/include)
    LIST(APPEND CMAKE_LIBRARY_PATH ${OSGEARTH_DEPENDENCY_DIR}/lib)
ENDIF()

**Comment by gwaldron** [comment]
Indeed, it works best with something like `vcpkg` that installs everything in the same place.

Per your original comment, the `OSG_DIR` variable will do exactly as you suggest -- just fill that in, and the rest of the OSG variables should populate automatically.

**Comment by JackBoosY** [comment]
Hi guys,
Recently I am trying to promote a PR of osgearth(https://github.com/microsoft/vcpkg/pull/15375) but met some problems:
1. The way osgearth finds dependencies does not support config mode (direct link to target).
2. In linux, the link order of the dependency osg is wrong, which leads to failure to link successfully.
The order should be `OSGPLUGIN1_LIBRARY OSGPLUGIN2_LIBRARY ... OSGUTIL_LIBRARY OSG_LIBRARY`.
Note that the symbols in the library in Linux are always searched behind the link list.

Can anyone help me?

Thanks.

**Reply by plevy** [reply]
@JackBoosY I have been looking at updating the vcpkg port to 3.2.  Do you still have any outstanding issues or requests that we should clean up for 3.2?

Thank you

**Reply by JackBoosY** [reply]
Currently no, thanks!

**Comment by gwaldron** [comment]
Closing - vcpkg distro is up to date
